File: backend/app/schemas/observability/insights.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from app.schemas.common import UTCModel


InsightsRange = Literal["24h", "7d", "30d", "90d"]
InsightsInterval = Literal["auto", "hour", "day"]
ResolvedInsightsInterval = Literal["hour", "day"]


class NamedMetric(UTCModel):
    name: str
    value: int | float


class TimeSeriesPoint(UTCModel):
    bucket_start_at: datetime
    value: int | float


class RankedCocoonMetric(UTCModel):
    cocoon_id: str
    cocoon_name: str
    value: int | float


class InsightsSummary(UTCModel):
    total_messages: int = 0
    total_runs: int = 0
    total_tokens: int = 0
    error_rate: float = 0
    average_latency_ms: float = 0
    active_cocoons: int = 0
    pending_wakeup_count: int = 0


class TokenUsageInsights(UTCModel):
    series: list[TimeSeriesPoint]
    by_provider: list[NamedMetric]
    by_model: list[NamedMetric]
    by_operation: list[NamedMetric]


class MemoryInsights(UTCModel):
    total_memories: int = 0
    growth: list[TimeSeriesPoint]
    by_source_kind: list[NamedMetric]
    by_memory_type: list[NamedMetric]
    top_cocoons: list[RankedCocoonMetric]


class RuntimeInsights(UTCModel):
    request_series: list[TimeSeriesPoint]
    decision_distribution: list[NamedMetric]
    status_distribution: list[NamedMetric]
    node_latency: list[NamedMetric]
    latency_p50_ms: float = 0
    latency_p95_ms: float = 0
    silence_rate: float = 0
    wakeup_rate: float = 0
    error_rate: float = 0
    top_error_cocoons: list[RankedCocoonMetric]


class InsightsDashboard(UTCModel):
    generated_at: datetime
    range: InsightsRange
    interval: ResolvedInsightsInterval
    summary: InsightsSummary
    token_usage: TokenUsageInsights
    memory: MemoryInsights
    runtime: RuntimeInsights

File: frontend/src/api/types/insights.ts
import type { NamedMetric, RankedCocoonMetric, TimeSeriesPoint } from "./common";

export type { NamedMetric, RankedCocoonMetric, TimeSeriesPoint } from "./common";

export type InsightsRange = "24h" | "7d" | "30d" | "90d";
export type InsightsInterval = "auto" | "hour" | "day";
export type ResolvedInsightsInterval = "hour" | "day";

export interface InsightsSummary {
  total_messages: number;
  total_runs: number;
  total_tokens: number;
  error_rate: number;
  average_latency_ms: number;
  active_cocoons: number;
  pending_wakeup_count: number;
}

export interface TokenUsageInsights {
  series: TimeSeriesPoint[];
  by_provider: NamedMetric[];
  by_model: NamedMetric[];
  by_operation: NamedMetric[];
}

export interface MemoryInsights {
  total_memories: number;
  growth: TimeSeriesPoint[];
  by_source_kind: NamedMetric[];
  by_memory_type: NamedMetric[];
  top_cocoons: RankedCocoonMetric[];
}

export interface RuntimeInsights {
  request_series: TimeSeriesPoint[];
  decision_distribution: NamedMetric[];
  status_distribution: NamedMetric[];
  node_latency: NamedMetric[];
  latency_p50_ms: number;
  latency_p95_ms: number;
  silence_rate: number;
  wakeup_rate: number;
  error_rate: number;
  top_error_cocoons: RankedCocoonMetric[];
}

export interface InsightsDashboard {
  generated_at: string;
  range: InsightsRange;
  interval: ResolvedInsightsInterval;
  summary: InsightsSummary;
  token_usage: TokenUsageInsights;
  memory: MemoryInsights;
  runtime: RuntimeInsights;
}
